### PR TITLE
US040 - Login System (Overhaul)

### PR DESCRIPTION
**Pull request by BlankAmber**
# Pull Request Checklist

Please follow all these rules before merging your branch.
- [x] Must reference a ticket
- [x] Must implement tests where appropirate
- [x] Must have useful documentation
- [x] Must adhere to the naming conventions
- [x] Must not commit confidential information 

Once all requirements above are fulfilled, your pull request has no merge conflicts, and you have two approvals from the approved approvers, then it can be tested and signed off.

# Relevant tickets
- Implements backend for [US040 - Login System](https://tasks.office.com/kainos.com/Home/Task/b_OhVXRyQEGETWG0--tRX5YAKUk2?Type=TaskLink&Channel=Link&CreatedTime=638382480756900000)
- Implements backend for [US041 - Authorisation System](https://tasks.office.com/kainos.com/Home/Task/1tQcqEQ2KkWmi3YhtxdI2JYAOcgr?Type=TaskLink&Channel=Link&CreatedTime=638382480898780000)

# Changes
- Replaced SHA-256 with Argon2 using [password4j](https://github.com/Password4j/password4j)
  - To generate the hashed passwords (should be implemented in [US024 - Registration System](https://tasks.office.com/kainos.com/Home/Task/-LpElUzK40epI97zNycNUZYABeGF?Type=TaskLink&Channel=Link&CreatedTime=638382482281550000)) use: `Password.hash(login.getPassword()).addRandomSalt(12).withArgon2()`
- Generates a JWT upon logging in using [auth0](https://github.com/auth0/java-jwt)
  - JWT_SECRET environment variable added
  - Secret generated using `node -e "console.log(require('crypto').randomBytes(256).toString('hex'));"`
- Implements authorisation
  - Note that POST /login does not accept a JWT query param as users should not be logged in beforehand
  - All future endpoints (excluding POST /register) should accept a JWT query param
  - The GET /all-job-roles endpoint now accepts a JWT query param
  - All future endpoints (excluding POST /register) should accept a JWT query param

# Screenshots
Screenshot 2023-12-15 at 18 34 26

JWT token decoded using [jwt.io](https://jwt.io/)
Screenshot 2023-12-15 at 18 34 49

Screenshot 2023-12-15 at 18 33 16

Screenshot 2023-12-15 at 18 33 43

# Future Considerations
- The hashed password is stored as a `varchar(100)` in the database, is this the most appropriate data type?
- SSO (Google?)
- Should POST /login still be POST since it's not inserting anything into the database?
- More unit tests? The controllers and DAOs could be unit tested

### CODE REVIEW

**Review by adammarshall745 — DISMISSED** [review]
Looks good! Tests run beautifully (:

File: src/main/java/org/kainos/ea/db/RoleID.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package org.kainos.ea.db;
+
+public final class RoleID {
+    private RoleID() {
+
+    }
+
+    public static final int ADMIN = 1;
+    public static final int EMPLOYEE = 2;
+}
```
Comment (jkelso662) [inline src/main/java/org/kainos/ea/db/RoleID.java:10]: Nearly! I think what you need an ENUM class :)

**Review by jkelso662 — CHANGES_REQUESTED** [review]
Really good! Great overhaul and very close to how we might implement this on project. Code also looks a lot cleaner using these other libraries! Left one comment for a class that needs fixing. Also seeing a few empty constructors. Make sure you ask the question, does it make sense here to create an empty object? 

One other thing to think about - looks like you intend on sending the JWT token in the URL as a query parameter. It is one way of doing it but might be more common to have it in the request body or more common than that - in an Auth header. I assume you'll store your token in your front end session object? Anyway - really good work. Yous a re flying through this!

**Review by adammarshall745 — APPROVED** [review]
Good improvements Samuel b. I have re run the tests and they passed fine

**Review by jkelso662 — APPROVED** [review]
Amazing! 🥇

**Pull request merged**